src/snarks/groth16/mpc/phase2.hpp

```diff
@@ -0,0 +1,309 @@
+#ifndef __ZETH_SNARKS_GROTH16_MPC_PHASE2_HPP__
+#define __ZETH_SNARKS_GROTH16_MPC_PHASE2_HPP__
+
+#include "include_libsnark.hpp"
+#include "snarks/groth16/mpc/hash_utils.hpp"
+
+// Structures and operations related to the "Phase 2" MPC described in
+// [BoweGM17].  Parts of the implementation use techniques from the
+// "Phase2" library from "zk-SNARK MPCs, made easy".
+//
+// References:
+//
+// \[BoweGM17]
+//  "Scalable Multi-party Computation for zk-SNARK Parameters in the Random
+//  Beacon Model"
+//  Sean Bowe and Ariel Gabizon and Ian Miers,
+//  IACR Cryptology ePrint Archive 2017,
+//  <http://eprint.iacr.org/2017/1050>
+//
+// "Phase2" (From "zk-SNARK MPCs, made easy" library
+//  https://github.com/ebfull/phase2)
+//
+// "Sapling MPC" ("Multi-party computation for Zcash's Sapling zk-SNARK public
+//  parameters" https://github.com/zcash-hackworks/sapling-mpc)
+namespace libzeth
+{
+
+template<typename ppT> class srs_powersoftau;
+template<typename ppT> class srs_mpc_layer_L1;
+
+/// Target of the MPC for Phase2 of the SRS generation.  Follows exactly $M_2$
+/// in section 7.3 of [BoweGM17], whre we use L in place of K, consistent with
+/// the keypair in libsnark.
+template<typename ppT> class srs_mpc_phase2_accumulator
+{
+public:
+    // Hash of the initial state before any contributions are made.  Kept
+    // constant over the MPC, and used to check that challenges and responses
+    // are part of the same MPC.
+    srs_mpc_hash_t cs_hash;
+
+    libff::G1<ppT> delta_g1;
+
+    libff::G2<ppT> delta_g2;
+
+    // { H_i } = { [ t(x) . x^i / delta ]_1 }  i=0..n-2 (n-1 entries)
```

Comment:
`i=0..n-2 (n-1 entries)` => `i \in [0..n-2] (n-1 entries)` likewise for `j` below

Reply:
changed